spring-web/src/main/java/org/springframework/web/multipart/reactive/nio/NioMultipartResolver.java

```diff
@@ -0,0 +1,167 @@
+ */
+public class NioMultipartResolver implements MultipartResolver {
+
+	@Override
+	public boolean isMultipart(ServerHttpRequest request) {
+		if (!HttpMethod.POST.equals(request.getMethod())) {
+			return false;
+		}
+		MediaType contentType = request.getHeaders().getContentType();
+		return (contentType != null && MultipartUtils.isMultipart(contentType.toString()));
+	}
+
+	@Override
+	public Flux<Part> resolveParts(ServerHttpRequest request) {
+		if (!isMultipart(request)) {
+			return Flux.empty();
+		}
+		return Flux.create(new NioMultipartConsumer(request));
+	}
+
+	@Override
+	public Mono<MultiValueMap<String, Part>> resolveAllParts(ServerHttpRequest request) {
+		return resolveParts(request)
+				.collectList()
+				.map(parts -> {
+					MultiValueMap<String, Part> partsMap = new LinkedMultiValueMap<>();
+					parts.forEach(part -> partsMap.add(part.getName(), part));
+					return partsMap;
+				});
+	}
+
+	private static class NioMultipartConsumer implements Consumer<FluxSink<Part>> {
+
+		private final ServerHttpRequest request;
+
+
+		public NioMultipartConsumer(ServerHttpRequest request) {
+			this.request = request;
+		}
+
+
+		@Override
+		public void accept(FluxSink<Part> emitter) {
+			HttpHeaders headers = request.getHeaders();
+			MultipartContext context = new MultipartContext(
+					headers.getContentType().toString(),
+					Math.toIntExact(headers.getContentLength()),
+					headers.getFirst(HttpHeaders.ACCEPT_CHARSET));
+			NioMultipartParser parser = Multipart.multipart(context).forNIO(new NioMultipartParserListener() {
+				@Override
+				public void onPartFinished(StreamStorage streamStorage, Map<String, List<String>> headersFromPart) {
+					HttpHeaders headers = new HttpHeaders();
+					headers.putAll(headersFromPart);
+					emitter.next(new NioPart(headers, streamStorage));
+				}
+
+				@Override
+				public void onFormFieldPartFinished(String fieldName, String fieldValue, Map<String, List<String>> headersFromPart) {
+					HttpHeaders headers = new HttpHeaders();
+					headers.putAll(headersFromPart);
+					emitter.next(new NioPart(headers, fieldValue));
+				}
+
+				@Override
+				public void onAllPartsFinished() {
+					emitter.complete();
+				}
+
+				@Override
+				public void onNestedPartStarted(Map<String, List<String>> headersFromParentPart) {
+				}
+
+				@Override
+				public void onNestedPartFinished() {
+				}
+
+				@Override
+				public void onError(String message, Throwable cause) {
+					emitter.error(new MultipartException(message, cause));
+				}
+			});
+
+			request.getBody().subscribe(buffer -> {
+				byte[] resultBytes = new byte[buffer.readableByteCount()];
+				buffer.read(resultBytes);
+				try {
+					parser.write(resultBytes);
+				}
+				catch (IOException ex) {
+					throw Exceptions.bubble(ex);
+				}
+
+			}, (e) -> {
+				try {
+					parser.close();
+				}
+				catch (IOException ex) {
+					throw Exceptions.bubble(e);
+				}
```

Comment:
Should this notify the parser about the error which calls `error` on the emitter, so that subscribers for the parts will get the error?